pgmpy/models/DoubleMLRegressor.py

```
@@ -0,0 +1,428 @@
+        """
+        if isinstance(X, pd.DataFrame):
+            df = X.copy()
+            if self.feature_names is not None:
+                # enforce caller-provided mapping order (user must ensure correct order)
+                if len(self.feature_names) != df.shape[1]:
+                    raise ValueError(
+                        "feature_names length does not match DataFrame columns. "
+                        f"{len(self.feature_names)} != {df.shape[1]}"
+                    )
+                df.columns = list(self.feature_names)
+            return df
+
+        # numpy array path
+        arr = np.asarray(X)
+        if arr.ndim == 1:
+            arr = arr.reshape(-1, 1)
+        n_cols = arr.shape[1]
+
+        if self.feature_names is not None:
+            if len(self.feature_names) != n_cols:
+                raise ValueError(
+                    "feature_names length does not match number of columns in X. "
+                    f"{len(self.feature_names)} != {n_cols}"
+                )
+            cols = list(self.feature_names)
+            return pd.DataFrame(arr, columns=cols)
+
+        if self.allow_array_unnamed:
+            warnings.warn(
+                "Passing a NumPy array without feature_names: columns will be named x0, x1, ... "
+                "This is unsafe for causal estimation; prefer passing a DataFrame or feature_names.",
+                UserWarning,
+            )
+            cols = [f"x{i}" for i in range(n_cols)]
+            return pd.DataFrame(arr, columns=cols)
+
+        # reject ambiguous arrays
+        raise ValueError(
+            "Ambiguous input: X is a NumPy array but no feature_names were provided. "
+            "Either pass X as a pandas DataFrame with column names matching the DAG roles, "
+            "or construct the estimator with feature_names=[...]."
+        )
+
+    def _read_roles(self) -> Tuple[str, List[str]]:
+        if not (
+            hasattr(self.dag, "get_role")
```

Quick suggestion: maybe instead of checking this you can directly implement the `get_role` function! Thoughts @ankurankan @marrov ?

---

@DARHWOLF, I agree with @XAheli  here - given that you are already forcing that `dag` should be a `pgmpy` DAG it is ok to directly use `get_role`. This method is essentially an undercorver type validation. Just let it fail loudly if the user does not input a `DAG`.